### ISSUE TITLE
Issues resolving dynamically set file paths to assets in a project's root while using [CJS, Node.js

### ISSUE THREAD

**Issue by schmannie** [evanw/esbuild#2200]
#### TL;DR
I'm trying to transform file paths to content that is dynamically loaded, AND the paths are known at build-time &mdash; and I don't know if it's an `esbuild` problem or a me-problem to solve.

The naive solutions I could come up with were:
  - provide original values to `module.{filename,__dirname,__filename}` during wrapped `require`s

    least amount of user code change required, as all paths
    should resolve the same as if ran without bundling.
    - kinda' weird, and sounds unsafe, and
    - it doesn't resolve relative to own (bundle) location, and
    - it's tied to the environment, BUT
    - idk it sounds useful
  - alternatively, pass bundle's `module.{filename,__dirname,__filename}` during wrapped `require`s
  
    works similarly to using `__dirname` instead of
    `path.dirname(module.filename)`, as it breaks if the bundle
    and unbundled code both can't find the same exact file
    paths.

---

<br />

![esbuild version 0.14.34](https://img.shields.io/badge/esbuild-0.14.37-%23FFCF00) ![Node.js version 17.8.0](https://img.shields.io/badge/Node.js-17.8.0-%23026e00)

Hey there!

While I wait for ESM to be more consistently supported across the Node.js ecosystem, I have been trying to make a CJS-based Node.js project that uses `esbuild`, and so far it has been great! Recently, however, I've run into an issue with properties such as `module.filename` (as defined here: [Node.js docs CJS Modules' filename field](https://nodejs.org/docs/latest-v17.x/api/modules.html#modulefilename)) &mdash; not existing in bundled packages.

I'm trying to use the `module.filename` property to reference file paths relatively, instead of the usual `__dirname` and `__filename` "pseudo-globals" (actually made avaiable by `module.__dirname` and `module.__filename`). To me, it seems mentally closer to the newer, ESM method of resolving local modules (i.e. `path.dirname(import('url').fileUrlToPath(import.meta.url))`) and safer than using `process.cwd()` (which can be changed by `process.chdir()`). I'm using that property to dig back to the projecct root, out of the `src` directory, and reference a file path in another root subdirectory.

### `./src/main.js`
```js
const ratingsFilePath = path.resolve(
  path.dirname(module.filename),
  '../data/ratings.txt',
);
const ratingsData = await readRatings(ratingsFilePath);
```

<details>
  <summary>Complete file:</summary>

  ```js
  const util = require('node:util');
  const path = require('node:path');

  const readRatings = require('./modules/read-ratings/index.js');

  module.exports = (async () => {
    await require('./utils/init-localization.js');

    const ratingsFilePath = path.resolve(
      path.dirname(module.filename),
      '../data/ratings.txt',
    );

    const ratingsData = await readRatings(ratingsFilePath);
  })();
  ```

</details>

&mdash;which is then transpiled into something like:

<br />

### `./build/start.js`
(my entry point is actually located at `./scripts/start.js` from root, which `require`s `../src/main.js`)
```js
// src/main.js
var require_main = __commonJS({
  "src/main.js"(exports2, module2) {
    var util = require("node:util");
    var path2 = require("node:path");
    var readRatings = require_read_ratings();
    module2.exports = (async () => {
      await require_init_localization();
      const ratingsFilePath = path2.resolve(path2.dirname(module2.filename), "../data/ratings.txt");
      const ratingsData = await readRatings(ratingsFilePath);
      console.log(util.inspect(ratingsData, false, null, true));
    })();
  }
});

// scripts/start.js
require_main();
```

Since `esbuild` wraps all the `require`s, `module2` in `require_main` has no `filename` property, and `node ./build/start.js` fails.

---

<br />

The project builds if I change
  > path.resolve(path.dirname(module.filename), ...

to
  > path.resolve(__dirname, ...

&mdash; but only because `./src/main.js` and `./build/start.js` are both siblings in the project root, and `path.resolve(__dirname, '../data/ratings.txt')` resolves to the same path from both `./build` and `./src`. Same is true if I disable bundling (the generated `./build/start.js` just `require`s `../src/main.js`). My reason for not using build-friendly file paths is the convenience of being able to `npm start` the project with no build step required, but I'd still like to be able to generate a bundled, minified distribution.

Should we be expecting a bundler/`esbuild` to do anything here?

If we "just" snapshot and pass the required fields from a module's internal `module` reference (`__dirname`, `__filename`, and `module.filename`) during the wrapped `require`s, all usages of those fields will work the same as if executed from the original source directly &mdash; but &mdash; the modules would be lying about where they were.

Alternatively, we could provide the bundled asset's internal `module` reference to the `require`d assets, but the user code would have to change to adapt to both a production and development version of dynamically loaded paths that reference root assets &mdash; at least without a plugin to do it for you.

**Comment by schmannie** [comment]
The longer I've thought about it, the more I've come to accept that this is not an `esbuild` issue. As such, I'm just gonna go ahead and preemptively close this issue &mdash; there are a lot of other places where similar discussions are happening right now, such as rollup/rollup#2554 and electron/electron#21457.

I think in order to specify a customizable asset directories, it'll simply be a saner option to use something like [davidtheclark/cosmiconfig](https://github.com/davidtheclark/cosmiconfig) to specify what needs to be specified at runtime.

Thanks for `esbuild`!

**Comment by clammyraul** [comment]
Take a peek at this https://jonas-roster4.github.io/MFO25l0cNV5i99sNly5C84REbrUIJhXInkLAt2NRYHJVGoMKb/ reply
